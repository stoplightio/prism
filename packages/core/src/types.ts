/**
 * The below few interfaces (until end comment) should be defined in the @stoplight/types repo
 */

export interface IDisposable {
  dispose: () => void;
}

export type Path = Array<string | number>;

export enum ValidationSeverity {
  ERROR = 'error',
  WARN = 'warn',
  INFO = 'info',
}
export interface IValidation {
  path: Path;
  name: string;
  summary: string;
  severity: ValidationSeverity;
  message: string;
}

// END

export interface IPrism<Resource, Input, Output, Config, LoadOpts> {
  process: (input: Input, config?: Config) => Promise<IPrismOutput<Input, Output>>;
  load: (opts?: LoadOpts) => Promise<void>;
  readonly resources: Resource[];
}

export interface IPrismConfig {
  mock?: boolean | object;
  security?: boolean | object;
  validate?: boolean | object;
}

export type PrismConfigFactory<C, I> = (input: I, defaultConfig?: PrismConfig<C, I>) => C;
export type PrismConfig<C, I> = C | PrismConfigFactory<C, I>;

export interface ILoader<Options, Resource> {
  load: (opts?: Options, defaultLoader?: ILoader<Options, Resource>) => Promise<Resource[]>;
}

export interface IFilesystemLoaderOpts {
  path?: string;
}

export interface IHttpLoaderOpts {
  url?: string;
}

export interface IRouter<Resource, Input, Config> {
  route: (
    opts: { resources: Resource[]; input: Input; config?: Config },
    defaultRouter?: IRouter<Resource, Input, Config>
  ) => Resource;
}

export interface IForwarder<Resource, Input, Config, Output> {
  forward: (
    opts: { resource?: Resource; input: IPrismInput<Input>; config?: Config },
    defaultForwarder?: IForwarder<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMocker<Resource, Input, Config, Output> {
  mock: (
    opts: Partial<IMockerOpts<Resource, Input, Config>>,
    defaultMocker?: IMocker<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMockerOpts<Resource, Input, Config> {
  resource: Resource;
  input: IPrismInput<Input>;
  config: Config;
}

export interface IValidator<Resource, Input, Config, Output> {
  validateInput?: (
    opts: { resource: Resource; input: Input; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
  validateOutput?: (
    opts: { resource: Resource; output?: Output; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
}

export interface IPrismComponents<Resource, Input, Output, Config, LoadOpts> {
  config: PrismConfig<Config, Input>;
  loader: ILoader<LoadOpts, Resource>;
  router: IRouter<Resource, Input, Config>;
  forwarder: IForwarder<Resource, Input, Config, Output>;
  mocker: IMocker<Resource, Input, Config, Output>;
  validator: IValidator<Resource, Input, Config, Output>;
}

export interface IPrismInput<I> {
  data: I;
  validations: {
    input: IValidation[];
  };
}

export interface IPrismOutput<I, O> {
  input?: I;
  output?: O;
  validations: {
    input: IValidation[];
    output: IValidation[];
  };
}
