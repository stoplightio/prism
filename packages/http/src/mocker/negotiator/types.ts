import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
