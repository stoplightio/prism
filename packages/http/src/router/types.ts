import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
