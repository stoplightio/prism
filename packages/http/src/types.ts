import { IPrism, IPrismComponents, IPrismConfig } from '@stoplight/prism-core';
import { IHttpOperation, Omit } from '@stoplight/types';

export type TPrismHttpInstance<LoaderInput> = IPrism<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  LoaderInput
>;

export type TPrismHttpComponents<LoaderInput> = Partial<
  IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, LoaderInput>
>;

// TODO: should be complete | and in the @stoplight/types repo
export type IHttpMethod = 'get' | 'put' | 'post' | 'delete' | 'options' | 'head' | 'patch' | 'trace'; // ... etc

export interface IHttpOperationConfig {
  mediaType?: string;
  code?: string;
  exampleKey?: string;
  dynamic: boolean;
}

export interface IHttpConfig extends IPrismConfig {
  mock: false | IHttpOperationConfig;

  validate?: {
    request?:
      | boolean
      | {
          hijack?: boolean;
          headers?: boolean;
          query?: boolean;
          body?: boolean;
        };

    response?:
      | boolean
      | {
          headers?: boolean;
          body?: boolean;
        };
  };
}

export interface IHttpNameValues {
  [name: string]: string | string[];
}

export interface IHttpNameValue {
  [name: string]: string;
}

export interface IHttpUrl {
  baseUrl?: string;
  path: string;
  query?: IHttpNameValues;
}

export interface IHttpRequest {
  method: IHttpMethod;
  url: IHttpUrl;
  headers?: IHttpNameValue;
  body?: any;
}

export interface IHttpResponse {
  statusCode: number;
  headers?: IHttpNameValue;
  body?: any;
}

export type ProblemJson = {
  type: string;
  title: string;
  status: number;
  detail: string;
};

export class ProblemJsonError extends Error {
  public static fromTemplate(template: Omit<ProblemJson, 'detail'>, detail?: string): ProblemJsonError {
    Error.captureStackTrace(this, ProblemJsonError);
    return new ProblemJsonError(
      `https://stoplight.io/prism/errors#${template.type}`,
      template.title,
      template.status,
      detail || '',
    );
  }

  public static fromPlainError(error: Error & { detail?: string; status?: number }): ProblemJson {
    return {
      type: error.name && error.name !== 'Error' ? error.name : 'https://stoplight.io/prism/errors#UNKNOWN',
      title: error.message,
      status: error.status || 500,
      detail: error.detail || '',
    };
  }

  constructor(readonly name: string, readonly message: string, readonly status: number, readonly detail: string) {
    super(message);
    Error.captureStackTrace(this, ProblemJsonError);
  }
}

export type PayloadGenerator = (f: unknown) => Promise<unknown>;
