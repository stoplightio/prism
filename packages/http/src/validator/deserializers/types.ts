import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValue, IHttpNameValues, JSONSchema } from '../../types';

export type deserializeFn<Parameters> = (
  name: string,
  parameters: Parameters,
  schema?: JSONSchema,
  explode?: boolean
) => unknown;

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: deserializeFn<Parameters>;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
