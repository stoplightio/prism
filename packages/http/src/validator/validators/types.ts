import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { JSONSchema } from 'http/src/types';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IPrismDiagnostic[];
}

export interface ISchemaValidator<S = JSONSchema> {
  validate(content: any, schema: S): IPrismDiagnostic[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: JSONSchema) => IPrismDiagnostic[]) | undefined;
}
