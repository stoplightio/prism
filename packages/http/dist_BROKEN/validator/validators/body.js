"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = exports.findContentByMediaTypeOrFirst = exports.decodeUriEntities = exports.splitUriParams = exports.deserializeFormBody = void 0;
const types_1 = require("@stoplight/types");
const A = require("fp-ts/Array");
const E = require("fp-ts/Either");
const O = require("fp-ts/Option");
const NEA = require("fp-ts/NonEmptyArray");
const function_1 = require("fp-ts/function");
const lodash_1 = require("lodash");
const type_is_1 = require("type-is");
const deserializers_1 = require("../deserializers");
const utils_1 = require("./utils");
const types_2 = require("./types");
const filterRequiredProperties_1 = require("../../utils/filterRequiredProperties");
function deserializeFormBody(schema, encodings, decodedUriParams) {
    if (!schema.properties) {
        return decodedUriParams;
    }
    return function_1.pipe(Object.keys(schema.properties), A.reduce({}, (deserialized, property) => {
        var _a;
        deserialized[property] = decodedUriParams[property];
        const encoding = encodings.find(enc => enc.property === property);
        if (encoding && encoding.style) {
            const deserializer = deserializers_1.body[encoding.style];
            const propertySchema = (_a = schema.properties) === null || _a === void 0 ? void 0 : _a[property];
            if (propertySchema && typeof propertySchema !== 'boolean')
                deserialized[property] = deserializer(property, decodedUriParams, propertySchema);
        }
        return deserialized;
    }));
}
exports.deserializeFormBody = deserializeFormBody;
function splitUriParams(target) {
    return target.split('&').reduce((result, pair) => {
        const [key, ...rest] = pair.split('=');
        result[key] = rest.join('=');
        return result;
    }, {});
}
exports.splitUriParams = splitUriParams;
function decodeUriEntities(target) {
    return Object.entries(target).reduce((result, [k, v]) => {
        result[decodeURIComponent(k)] = decodeURIComponent(v);
        return result;
    }, {});
}
exports.decodeUriEntities = decodeUriEntities;
function findContentByMediaTypeOrFirst(specs, mediaType) {
    return function_1.pipe(specs, A.findFirst(spec => !!type_is_1.is(mediaType, [spec.mediaType])), O.alt(() => A.head(specs)), O.map(content => ({ mediaType, content })));
}
exports.findContentByMediaTypeOrFirst = findContentByMediaTypeOrFirst;
function deserializeAndValidate(content, schema, target, bundle) {
    const encodings = lodash_1.get(content, 'encodings', []);
    const encodedUriParams = splitUriParams(target);
    return function_1.pipe(validateAgainstReservedCharacters(encodedUriParams, encodings), E.map(decodeUriEntities), E.map(decodedUriEntities => deserializeFormBody(schema, encodings, decodedUriEntities)), E.chain(deserialised => function_1.pipe(utils_1.validateAgainstSchema(deserialised, schema, true, undefined, bundle), E.fromOption(() => deserialised), E.swap)));
}
function memoizeSchemaNormalizer(normalizer) {
    const cache = new WeakMap();
    return (schema) => {
        const cached = cache.get(schema);
        if (!cached) {
            const newSchema = normalizer(schema);
            cache.set(schema, newSchema);
            return newSchema;
        }
        return cached;
    };
}
const normalizeSchemaProcessorMap = {
    [types_2.ValidationContext.Input]: memoizeSchemaNormalizer(filterRequiredProperties_1.stripReadOnlyProperties),
    [types_2.ValidationContext.Output]: memoizeSchemaNormalizer(filterRequiredProperties_1.stripWriteOnlyProperties),
};
const validate = (target, specs, context, mediaType, bundle) => {
    const findContentByMediaType = function_1.pipe(O.Do, O.bind('mediaType', () => O.fromNullable(mediaType)), O.bind('contentResult', ({ mediaType }) => findContentByMediaTypeOrFirst(specs, mediaType)), O.alt(() => O.some({ contentResult: { content: specs[0] || {}, mediaType: 'random' } })), O.bind('schema', ({ contentResult }) => function_1.pipe(O.fromNullable(contentResult.content.schema), O.chain(normalizeSchemaProcessorMap[context]))));
    return function_1.pipe(findContentByMediaType, O.fold(() => E.right(target), ({ contentResult: { content, mediaType: mt }, schema }) => function_1.pipe(mt, O.fromPredicate(mediaType => !!type_is_1.is(mediaType, ['application/x-www-form-urlencoded'])), O.fold(() => function_1.pipe(utils_1.validateAgainstSchema(target, schema, false, undefined, bundle), E.fromOption(() => target), E.swap), () => function_1.pipe(target, E.fromPredicate((target) => typeof target === 'string', () => [{ message: 'Target is not a string', code: '422', severity: types_1.DiagnosticSeverity.Error }]), E.chain(target => deserializeAndValidate(content, schema, target)))), E.mapLeft(diagnostics => applyPrefix('body', diagnostics)))));
};
exports.validate = validate;
function applyPrefix(prefix, diagnostics) {
    return function_1.pipe(diagnostics, NEA.map(d => ({ ...d, path: [prefix, ...(d.path || [])] })));
}
function validateAgainstReservedCharacters(encodedUriParams, encodings) {
    return function_1.pipe(encodings, A.reduce([], (diagnostics, encoding) => {
        const allowReserved = lodash_1.get(encoding, 'allowReserved', false);
        const property = encoding.property;
        const value = encodedUriParams[property];
        if (!allowReserved && /[/?#[\]@!$&'()*+,;=]/.test(value)) {
            diagnostics.push({
                path: [property],
                message: 'Reserved characters used in request body',
                severity: types_1.DiagnosticSeverity.Error,
            });
        }
        return diagnostics;
    }), diagnostics => (A.isNonEmpty(diagnostics) ? E.left(diagnostics) : E.right(encodedUriParams)));
}
