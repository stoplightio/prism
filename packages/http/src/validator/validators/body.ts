import { IMediaTypeContent } from '@stoplight/types';

import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { IHttpValidator, IValidatorRegistry } from './types';

export class HttpBodyValidator implements IHttpValidator<any, IMediaTypeContent> {
  constructor(private _registry: IValidatorRegistry, private _prefix: string) {}

  public validate(target: any, specs: IMediaTypeContent[], mediaType?: string): IPrismDiagnostic[] {
    const { _registry: registry, _prefix: prefix } = this;
    const content = this.getContent(specs, mediaType);

    if (!content) {
      return [];
    }

    if (!content.schema) {
      return [];
    }

    const validate = registry.get(content.mediaType);

    if (!validate) {
      return [];
    }

    return validate(target, content.schema).map(error =>
      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
    );
  }

  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
    if (!mediaType) {
      return specs[0];
    }

    const content = specs.find(c => c.mediaType === mediaType);

    if (!content) {
      return specs[0];
    }

    return content;
  }
}
