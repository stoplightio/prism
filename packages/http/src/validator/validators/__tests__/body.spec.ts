import { HttpParamStyles } from '@stoplight/types';
import { JSONSchema } from '../../..';
import { HttpBodyValidator } from '../body';

describe('HttpBodyValidator', () => {
  const httpBodyValidator = new HttpBodyValidator('body');

  describe('validate()', () => {
    describe('content specs are missing', () => {
      it('returns no validation errors', () => {
        expect(httpBodyValidator.validate('test', [])).toEqual([]);
      });
    });

    describe('request media type is not provided', () => {
      it('returns no validation errors', () => {
        expect(
          httpBodyValidator.validate('test', [
            { mediaType: 'application/not-exists-son', examples: [], encodings: [] },
          ]),
        ).toEqual([]);
      });
    });

    describe('request media type was not found in spec', () => {
      it('returns no validation errors', () => {
        expect(
          httpBodyValidator.validate(
            'test',
            [{ mediaType: 'application/not-exists-son', examples: [], encodings: [] }],
            'application/json',
          ),
        ).toEqual([]);
      });
    });

    describe('body schema is provided', () => {
      it('return validation errors', () => {
        const mockSchema: JSONSchema = { type: 'number' };
        expect(
          httpBodyValidator.validate(
            'test',
            [{ mediaType: 'application/json', schema: mockSchema, examples: [], encodings: [] }],
            'application/json',
          ),
        ).toMatchSnapshot();
      });
    });

    describe('body is form-urlencoded with deep object style', () => {
      it('returns no validation errors', () => {
        expect(
          httpBodyValidator.validate(
            encodeURI('key[a]=str'),
            [
              {
                mediaType: 'application/x-www-form-urlencoded',
                encodings: [{ property: 'key', style: HttpParamStyles.DeepObject }],
                schema: {
                  type: 'object',
                  properties: {
                    key: {
                      type: 'object',
                      properties: { a: { type: 'string' } },
                      required: ['a'],
                    },
                  },
                  required: ['key'],
                },
              },
            ],
            'application/x-www-form-urlencoded',
          ),
        ).toEqual([]);
      });
    });

    describe('body is form-urlencoded with deep object style and is not compatible with schema', () => {
      it('returns validation errors', () => {
        expect(
          httpBodyValidator.validate(
            encodeURI('key[a][ab]=str'),
            [
              {
                mediaType: 'application/x-www-form-urlencoded',
                encodings: [{ property: 'key', style: HttpParamStyles.DeepObject }],
                schema: {
                  type: 'object',
                  properties: {
                    key: {
                      type: 'object',
                      properties: {
                        a: {
                          type: 'object',
                          properties: { aa: { type: 'string' } },
                          required: ['aa'],
                        },
                      },
                      required: ['a'],
                    },
                  },
                  required: ['key'],
                },
              },
            ],
            'application/x-www-form-urlencoded',
          ),
        ).toMatchSnapshot();
      });
    });
  });
});
