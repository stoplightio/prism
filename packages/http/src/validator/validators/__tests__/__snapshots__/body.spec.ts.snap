// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpBodyValidator validate() body is form-urlencoded with deep object style and is not compatible with schema returns validation errors 1`] = `
Array [
  Object {
    "code": "required",
    "message": "should have required property 'aa'",
    "path": Array [
      "body",
      "key",
      "a",
    ],
    "severity": 0,
  },
]
`;

exports[`HttpBodyValidator validate() body schema is provided return validation errors 1`] = `
Array [
  Object {
    "code": "type",
    "message": "should be number",
    "path": Array [
      "body",
    ],
    "severity": 0,
  },
]
`;
